src/tests/stinger_batch_test/stinger_batch_test.cpp

```diff
@@ -0,0 +1,254 @@
+    stinger_batch_incr_edges<stinger_edge_update_adapter>(S, updates.begin(), updates.end());
+
+    consistency = stinger_consistency_check(S,S->max_nv);
+    EXPECT_EQ(consistency,0);
+
+    // Make sure all edges were detected as "new"
+    for (update_iterator u = updates.begin(); u != updates.end(); ++u)
+    {
+        EXPECT_EQ(u->result, 1);
+    }
+
+    // Reset return codes so we can reuse this batch
+    OMP("omp parallel for")
+    for (update_iterator u = updates.begin(); u < updates.end(); ++u) { u->result = 0; }
+
+    // Insert same batch again
+    stinger_batch_incr_edges<stinger_edge_update_adapter>(S, updates.begin(), updates.end());
+
+    consistency = stinger_consistency_check(S,S->max_nv);
+    EXPECT_EQ(consistency,0);
+
+    // Make sure all edges were detected as "updated"
+    for (update_iterator u = updates.begin(); u != updates.end(); ++u)
+    {
+        EXPECT_EQ(u->result, 0);
+    }
+
+    // Make sure all edge weights were incremented properly
+    STINGER_FORALL_EDGES_OF_ALL_TYPES_BEGIN(S) {
+        EXPECT_EQ(STINGER_EDGE_WEIGHT, 2);
+    }STINGER_FORALL_EDGES_OF_ALL_TYPES_END();
+
+    for (int i=0; i < 100; i++) {
+        EXPECT_EQ(stinger_outdegree_get(S, i), 99 - i);
+    }
+}
+
+TEST_F(StingerBatchTest, duplicate_batch_insertion) {
+    int64_t total_edges = 0;
+    int64_t consistency;
+
+    // Create batch to insert
+    std::vector<stinger_edge_update> updates;
+
+    const int num_dupes = 10;
+    int total_weight_per_edge = 0;
+    int64_t num_edges;
+    for (int d=0; d < num_dupes; ++d) {
+        num_edges = 0;
+        for (int i=0; i < 100; i++) {
+            for (int j=i+1; j < 100; j++) {
+                stinger_edge_update u = {
+                    0, NULL, // type, type_str
+                    i, NULL, // source, source_str
+                    j, NULL, // destination, destination_str
+                    d, // weight
+                    d, // time
+                    0, 0 // result, meta_index
+                };
+                updates.push_back(u);
+                num_edges++;
+            }
+        }
+        total_weight_per_edge += d;
+    }
+
+    // Do the updates
+    stinger_batch_incr_edges<stinger_edge_update_adapter>(S, updates.begin(), updates.end());
+
+    consistency = stinger_consistency_check(S,S->max_nv);
+    EXPECT_EQ(consistency,0);
+
+    // Only one update for each edge should return as "added", the rest should indicate "updated"
+    int64_t num_inserts = 0;
+    for (update_iterator u = updates.begin(); u < updates.end(); ++u)
+    {
+        if (u->result == 1) { ++num_inserts; }
+    }
+    EXPECT_EQ(num_inserts, num_edges);
+
+    // Weight should be sum of all updates
+    STINGER_FORALL_EDGES_OF_ALL_TYPES_BEGIN(S) {
+        EXPECT_EQ(STINGER_EDGE_WEIGHT, total_weight_per_edge);
+    }STINGER_FORALL_EDGES_OF_ALL_TYPES_END();
+
+    for (int i=0; i < 100; i++) {
+        EXPECT_EQ(stinger_outdegree_get(S, i), 99 - i);
+    }
+}
+
```

Comment:
I've decided to omit concurrent deletion tests for now. OpenMP won't let me call `stinger_batch_incr_edges` in parallel with `stinger_remove_edge`, because this would result in a `omp parallel for` nested inside an `omp section`. To fix this, `stinger_batch` functions should handle insertions and deletions together. While this is conceptually possible, it would be quite a bit more work, which doesn't seem worthwhile at the moment because concurrent insertion/deletion isn't actually being used anywhere at the moment. Let's just get this merged in, and we can create an issue to integrate optimized batch deletion.